src/MongoDB.Driver.Core/Core/ConnectionPools/ExclusiveConnectionPool.cs

```diff
@@ -712,95 +712,70 @@ public void Dispose()
+            public int Count => _connections.Count;
 
             public void Clear()
             {
-                lock (_lock)
+                PooledConnection connection;
+                while (_connections.TryAcquire(out connection))
                 {
-                    foreach (var connection in _connections)
-                    {
-                        RemoveConnection(connection);
-                    }
-                    _connections.Clear();
+                    DisposeConnection(connection);
                 }
             }
 
             public void Prune()
             {
-                lock (_lock)
+                PooledConnection expiredConnection;
+                while (_connections.TryRemoveMatchingLeastRecentlyUsed(c => c.IsExpired, out expiredConnection))
                 {
-                    for (int i = 0; i < _connections.Count; i++)
-                    {
-                        if (_connections[i].IsExpired)
-                        {
-                            RemoveConnection(_connections[i]);
-                            _connections.RemoveAt(i);
-                            break;
-                        }
-                    }
+                    DisposeConnection(expiredConnection);
                 }
             }
 
             public PooledConnection Acquire()
             {
-                lock (_lock)
+                PooledConnection connection;
+                if (_connections.TryAcquire(out connection))
                 {
-                    if (_connections.Count > 0)
+                    if (connection.IsExpired)
                     {
-                        var connection = _connections[_connections.Count - 1];
-                        _connections.RemoveAt(_connections.Count - 1);
-                        if (connection.IsExpired)
-                        {
-                            RemoveConnection(connection);
-                        }
-                        else
-                        {
-                            return connection;
-                        }
+                        DisposeConnection(connection);
+                    }
+                    else
+                    {
+                        return connection;
                     }
                 }
+
                 return null;
             }
 
             public void Return(PooledConnection connection)
             {
                 if (connection.IsExpired)
                 {
-                    RemoveConnection(connection);
+                    DisposeConnection(connection);
                     return;
                 }
 
-                lock (_lock)
-                {
-                    _connections.Add(connection);
-                }
+                _connections.Release(connection);
             }
 
-            private void RemoveConnection(PooledConnection connection)
+            // private methods
+            private void DisposeConnection(PooledConnection connection)
```

Comment:
I was having trouble with the name of this method. I concluded that the main point of this method is to Dispose of the connection (while raising events around that).